File: matlab/gtsam_examples/LocalizationExample.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% GTSAM Copyright 2010, Georgia Tech Research Corporation, 
% Atlanta, Georgia 30332-0415
% All Rights Reserved
% Authors: Frank Dellaert, et al. (see THANKS for the full author list)
% 
% See LICENSE for the license information
%
% @brief Example of a simple 2D localization example
% @author Frank Dellaert
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

import gtsam.*

%% Assumptions
%  - Robot poses are facing along the X axis (horizontal, to the right in 2D)
%  - The robot moves 2 meters each step
%  - The robot is on a grid, moving 2 meters each step

%% Create the graph (defined in pose2SLAM.h, derived from NonlinearFactorGraph)
graph = NonlinearFactorGraph;

%% Add two odometry factors
odometry = Pose2(2.0, 0.0, 0.0); % create a measurement for both factors (the same in this case)
odometryNoise = noiseModel.Diagonal.Sigmas([0.2; 0.2; 0.1]); % 20cm std on x,y, 0.1 rad on theta
graph.add(BetweenFactorPose2(1, 2, odometry, odometryNoise));
graph.add(BetweenFactorPose2(2, 3, odometry, odometryNoise));

%% Add three "GPS" measurements
% We use Pose2 Priors here with high variance on theta
priorNoise = noiseModel.Diagonal.Sigmas([0.1; 0.1; 10]);
graph.add(PriorFactorPose2(1, Pose2(0.0, 0.0, 0.0), priorNoise));
graph.add(PriorFactorPose2(2, Pose2(2.0, 0.0, 0.0), priorNoise));
graph.add(PriorFactorPose2(3, Pose2(4.0, 0.0, 0.0), priorNoise));

%% print
graph.print(sprintf('\nFactor graph:\n'));

%% Initialize to noisy points
initialEstimate = Values;
initialEstimate.insert(1, Pose2(0.5, 0.0, 0.2));
initialEstimate.insert(2, Pose2(2.3, 0.1,-0.2));
initialEstimate.insert(3, Pose2(4.1, 0.1, 0.1));
initialEstimate.print(sprintf('\nInitial estimate:\n  '));

%% Optimize using Levenberg-Marquardt optimization with an ordering from colamd
optimizer = LevenbergMarquardtOptimizer(graph, initialEstimate);
result = optimizer.optimizeSafely();
result.print(sprintf('\nFinal result:\n  '));

%% Plot trajectory and covariance ellipses
cla;
hold on;

plot2DTrajectory(result, [], Marginals(graph, result));

axis([-0.6 4.8 -1 1])
axis equal
view(2)

File: python/gtsam_unstable/examples/LocalizationExample.py
"""
A simple 2D pose slam example with "GPS" measurements
  - The robot moves forward 2 meter each iteration
  - The robot initially faces along the X axis (horizontal, to the right in 2D)
  - We have full odometry between pose
  - We have "GPS-like" measurements implemented with a custom factor
"""
import numpy as np

import gtsam
from gtsam import BetweenFactorPose2, Pose2, noiseModel
from gtsam_unstable import PartialPriorFactorPose2


def main():
    # 1. Create a factor graph container and add factors to it.
    graph = gtsam.NonlinearFactorGraph()

    # 2a. Add odometry factors
    # For simplicity, we will use the same noise model for each odometry factor
    odometryNoise = noiseModel.Diagonal.Sigmas(np.asarray([0.2, 0.2, 0.1]))

    # Create odometry (Between) factors between consecutive poses
    graph.push_back(
        BetweenFactorPose2(1, 2, Pose2(2.0, 0.0, 0.0), odometryNoise))
    graph.push_back(
        BetweenFactorPose2(2, 3, Pose2(2.0, 0.0, 0.0), odometryNoise))

    # 2b. Add "GPS-like" measurements
    # We will use PartialPrior factor for this.
    unaryNoise = noiseModel.Diagonal.Sigmas(np.array([0.1,
                                                      0.1]))  # 10cm std on x,y

    graph.push_back(
        PartialPriorFactorPose2(1, [0, 1], np.asarray([0.0, 0.0]), unaryNoise))
    graph.push_back(
        PartialPriorFactorPose2(2, [0, 1], np.asarray([2.0, 0.0]), unaryNoise))
    graph.push_back(
        PartialPriorFactorPose2(3, [0, 1], np.asarray([4.0, 0.0]), unaryNoise))
    graph.print("\nFactor Graph:\n")

    # 3. Create the data structure to hold the initialEstimate estimate to the solution
    # For illustrative purposes, these have been deliberately set to incorrect values
    initialEstimate = gtsam.Values()
    initialEstimate.insert(1, Pose2(0.5, 0.0, 0.2))
    initialEstimate.insert(2, Pose2(2.3, 0.1, -0.2))
    initialEstimate.insert(3, Pose2(4.1, 0.1, 0.1))
    initialEstimate.print("\nInitial Estimate:\n")

    # 4. Optimize using Levenberg-Marquardt optimization. The optimizer
    # accepts an optional set of configuration parameters, controlling
    # things like convergence criteria, the type of linear system solver
    # to use, and the amount of information displayed during optimization.
    # Here we will use the default set of parameters.  See the
    # documentation for the full set of parameters.
    optimizer = gtsam.LevenbergMarquardtOptimizer(graph, initialEstimate)
    result = optimizer.optimize()
    result.print("Final Result:\n")

    # 5. Calculate and print marginal covariances for all variables
    marginals = gtsam.Marginals(graph, result)
    print("x1 covariance:\n", marginals.marginalCovariance(1))
    print("x2 covariance:\n", marginals.marginalCovariance(2))
    print("x3 covariance:\n", marginals.marginalCovariance(3))


if __name__ == "__main__":
    main()
